### ISSUE TITLE
YAML Configuration | ATMS

### ISSUE THREAD

**Issue by danholdaway** [GEOS-ESM/swell#63]
## Agreement with GSI using GeoVaLs in UFO (Hofx no bias correction)

- [x] NOAA-20
- [x] NPP

## Agreement with GSI using GeoVaLs in UFO (Bias correction)

- [x] NOAA-20
- [x] NPP

## Agreement with GSI using GeoVaLs in UFO (Quality control)

- [x] NOAA-20
- [x] NPP

## Agreement with GSI using GeoVaLs in UFO (Hofx with bias correction)

- [x] NOAA-20
- [x] NPP

## Agreement with GSI using Swell

- [ ] NOAA-20
- [ ] NPP

## Waiting on JEDI PRs:

- [x] https://github.com/JCSDA-internal/crtm/pull/371
- [x] https://github.com/JCSDA-internal/ufo/pull/2162

**Comment by gmao-jjin3** [comment]
Here is an old pull request fixing the segment fault in ATMS emissivity over snow surface. 
https://github.com/JCSDA-internal/crtm-old/pull/74

**Comment by gmao-jjin3** [comment]
Here is the issue
https://github.com/JCSDA-internal/ufo-old/issues/1100

**Comment by gmao-jjin3** [comment]
The fix is in the UFO/CRTM and my previous test also passed with this fix in Sep 2020.

**Comment by gmao-jjin3** [comment]
Emily reported an issue about ATMS sea ice emissivity in Jan 2021. 
https://github.com/JCSDA-internal/ufo/issues/621

**Comment by gmao-yzhu** [comment]
We encountered an issue with error message "Caught SIGFPE: floating-point invalid operation" when calculating simulated ATMS radiances.

Similar to what Emily did, when I added a bunch of print statements, it ran successfully to the end for ATMS_n20 together with bias correction. the results are below:
27: Test : Vector difference between reference and computed: atms_n20 nobs= 232417 Min=-7.5294327245956083e-05, Max=7.5534442942171154e-05, RMS=9.6457825732815980e-06

However, it still failed for atms_npp at the first step -- **radiance simulation step** with error message "floating-point invalid operation", the detailed error message is copy-and-paste below:
**26: Caught SIGFPE: floating-point invalid operation**
26:  0# sigfpe_handler(int, siginfo_t*, void*) in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/liboops.so
26:  1# 0x00002AAAB7DC6C10 in /lib64/libpthread.so.0
26:  2# 0x00000000004E8E03 in /discover/nobackup/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/test_ObsOperator.x
26:  3# __svml_log2_l9 in /discover/nobackup/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/test_ObsOperator.x
**26:  4# nesdis_atms_seaice_module_mp_atms_seaice_bytbts_d_** in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26:  5# nesdis_atms_seaice_module_mp_nesdis_atms_seaice_ in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26:  6# crtm_mw_ice_sfcoptics_mp_compute_mw_ice_sfcoptics_ in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26:  7# crtm_sfcoptics_mp_crtm_compute_sfcoptics_ in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26:  8# common_rtsolution_mp_assign_common_input_ in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26:  9# crtm_rtsolution_mp_crtm_compute_rtsolution_ in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26: 10# 0x00002AAAB8880695 in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
26: 11# crtm_forward_module_mp_crtm_forward_ in /gpfsm/dnb32/yzhu/JEDI/fv3-bundle/build-intel-impi-release-fv3/bin/../lib64/libcrtm.so
``


Reported the issue to Ben Johnson at JCSDA.

**Comment by gmao-jjin3** [comment]
@gmao-yzhu  New testing files are in /discover/nobackup/jjin3/jedi/For_others/x0044.jj_20220619/

**Comment by danholdaway** [comment]
@gmao-yzhu the crash for NPP seems to be caused by missing data in the files. There are several locations where not all channels have a valid brightness temperature. For example, here are all the brightness temperatures at the crashing profile, as written out by the UFO (before the CRTM is called):

```
 -3.334767057904812E+038 -3.334767057904812E+038 -3.334767057904812E+038
 -3.334767057904812E+038 -3.334767057904812E+038 -3.334767057904812E+038
 -3.334767057904812E+038 -3.334767057904812E+038 -3.334767057904812E+038
 -3.334767057904812E+038 -3.334767057904812E+038 -3.334767057904812E+038
 -3.334767057904812E+038 -3.334767057904812E+038 -3.334767057904812E+038
 -3.334767057904812E+038   239.110000610352        254.279998779297     
   251.399993896484        247.970001220703        243.190002441406     
   237.839996337891
```

Then [here](https://github.com/JCSDA-internal/crtm/blob/944148bce060be7dbbe5a8a89e6dd06cdcbd1108/libsrc/NESDIS_ATMS_SeaICE_Module.f90#L225) there is a function that takes a log of these very large negative values. I strongly suspect that this is a big problem. It's not a bug, which makes me wonder why it works with the extra print statements you put in. So there might be other issues but this would be the first thing to resolve. 

It's not obvious to me how to solve this. Perhaps in the GSI these missing values are represented by very large _positive_ values which prevents the log from failing and then the points are filtered out later. Adjusting the value used for missing data might have unpredictable ramifications. This might need to be discussed with the JCSDA folks in order to come up with a workable solution.

**Comment by danholdaway** [comment]
You can run with this branch: https://github.com/JCSDA-internal/crtm/pull/371

**Comment by gmao-yzhu** [comment]
In addition to the AMSU-A like channels, ATMS has 6 more higher frequency channels including water vapor channels. I have made QC code changes for these higher frequencies and they are already merged into fv3-bundle. 

A bug is found and fixed in fv3-bundle hydrometeor code for ATMS. Will submit a PR to fv3-bundle. 

With the fix of the bug, consistent UFO results are obtained against GSI results for ATMS NOAA20. The comparisons of OmF with bias correction, histogram of OmF for data passing QC, final observation errors are shown below.  
![atms_n20 UFO_ObsValue-hofx ObsValue-GsiHofXBc_BC_QC ch17 ps](https://user-images.githubusercontent.com/73974278/176502903-2f4c5338-b22a-48d8-b4fc-f03600cc4650.png)
![atms_n20 UFO_Obs_Number O-F_BC_QC ch17 ps](https://user-images.githubusercontent.com/73974278/176502976-5675cc29-7ed2-4bb2-abdb-69433f71f1d0.png)
![atms_n20 UFO_EffectiveError GsiFinalObsError_QC ch17 ps](https://user-images.githubusercontent.com/73974278/176502996-22760d2c-f4e3-486b-b4d2-9c29b6ce7721.png)

**Comment by gmao-yzhu** [comment]
> [JCSDA-internal/crtm#371](https://github.com/JCSDA-internal/crtm/pull/371)

Thanks Dan for the crtm branch. I will try it for atms npp.

**Comment by danholdaway** [comment]
Great that everything works now for N20!

**Comment by gmao-jjin3** [comment]
@gmao-yzhu Can you check those observational errors in GSI and UFO as well?  Just a double-check, even though the histograms suggest they match.

**Comment by gmao-yzhu** [comment]
@gmao_jjin3 Yes, I had checked ATMS N20 observational errors in GSI and UFO, please see the above third figure. They matched very well. 

For ATMS NPP, I tested it in UFO with Dan's CRTM branch due to the CRTM sea ice issue, while the GSI results were still created using the original CRTM. 
The results looked good in terms of histogram of OmF passing QC, OmF with bias correction, and final observation errors.
The results for channel 5 and 18 are given below. For channel 5, the count of data passing QC in UFO (5068) is one less than that in GSI, but I think it is ok, given that different CRTM was used.  Channel 18 has the same number of data passing QC. 
![atms_npp UFO_Obs_Number O-F_BC_QC ch05](https://user-images.githubusercontent.com/73974278/176557522-13c1dbcd-a67d-4665-84a2-e73249109528.png)
![atms_npp UFO_Obs_Number O-F_BC_QC ch18](https://user-images.githubusercontent.com/73974278/176557536-5227c321-21be-4b51-8d14-4dcdd56185fa.png)
![atms_npp UFO_ObsValue-hofx ObsValue-GsiHofXBc_BC_QC ch05](https://user-images.githubusercontent.com/73974278/176557555-a871ab4d-9840-43e8-80c6-a72ab97c0ebe.png)
![atms_npp UFO_ObsValue-hofx ObsValue-GsiHofXBc_BC_QC ch18](https://user-images.githubusercontent.com/73974278/176557585-737f1594-f217-49ae-8234-1f90c28b76dc.png)
![atms_npp UFO_EffectiveError GsiFinalObsError_QC ch05](https://user-images.githubusercontent.com/73974278/176557601-ee0582bd-1acc-4627-85d5-735ccbc1e995.png)
![atms_npp UFO_EffectiveError GsiFinalObsError_QC ch18](https://user-images.githubusercontent.com/73974278/176557611-d6be3b7b-b47c-40e8-9791-d6cc25163989.png)

So the UFO tests of ATMS are completed now.

**Comment by danholdaway** [comment]
Great! Glad that the crtm changes are working out. Hopefully they will be merged soon.

**Comment by danholdaway** [comment]
@gmao-yzhu: the CRTM PR was merged so this should be working with the develop version of JEDI now. Tomorrow I will work on a new build of JEDI and then update the swell tag to point to a new build.
